remote-wallet/src/locator.rs

```diff
@@ -0,0 +1,628 @@
+            .unwrap()
+            .try_path(path)
+            .unwrap()
+            .try_query(maybe_query2)
+            .unwrap();
+
+        let uri = builder.build().unwrap();
+        write!(f, "{}", uri)
+    }
+}
+
+impl Locator {
+    pub fn new_from_path<P: AsRef<str>>(path: P) -> Result<Self, LocatorError> {
+        let path = path.as_ref();
+        let uri = URIReference::try_from(path)?;
+        Self::new_from_uri(&uri)
+    }
+
+    pub fn new_from_uri(uri: &URIReference<'_>) -> Result<Self, LocatorError> {
+        let scheme = uri.scheme().map(|s| s.as_str().to_ascii_lowercase());
+        let host = uri.host().map(|h| h.to_string());
+        match (scheme, host) {
+            (Some(scheme), Some(host)) if scheme == "usb" => {
+                let path = uri.path().segments().get(0).and_then(|s| {
+                    if !s.is_empty() {
+                        Some(s.as_str())
+                    } else {
+                        None
+                    }
+                });
+                let key = if let Some(query) = uri.query() {
+                    let query_str = query.as_str();
+                    let query = qstring::QString::from(query_str);
+                    if query.len() > 1 {
+                        return Err(DerivationPathError::InvalidDerivationPath(
+                            "invalid query string, extra fields not supported".to_string(),
+                        )
+                        .into());
+                    }
+                    let key = query.get("key");
+                    if key.is_none() {
+                        return Err(DerivationPathError::InvalidDerivationPath(format!(
+                            "invalid query string `{}`, only `key` supported",
+                            query_str,
+                        ))
+                        .into());
+                    }
+                    key.map(|v| v.to_string())
+                } else {
+                    None
+                };
+                Self::new_from_parts(host.as_str(), path, key.as_deref())
+            }
+            (Some(_scheme), Some(_host)) => Err(LocatorError::UnimplementedScheme),
+            (None, Some(_host)) => Err(LocatorError::UnimplementedScheme),
+            (_, None) => Err(LocatorError::ManufacturerError(ManufacturerError)),
+        }
+    }
+
+    pub fn new_from_parts<V, VE, P, PE, D, DE>(
+        manufacturer: V,
+        pubkey: Option<P>,
+        derivation_path: Option<D>,
+    ) -> Result<Self, LocatorError>
+    where
+        VE: Into<LocatorError>,
+        V: TryInto<Manufacturer, Error = VE>,
+        PE: Into<LocatorError>,
+        P: TryInto<Pubkey, Error = PE>,
+        DE: Into<LocatorError>,
+        D: TryInto<DerivationPath, Error = DE>,
+    {
+        let manufacturer = manufacturer.try_into().map_err(|e| e.into())?;
+        let pubkey = if let Some(pubkeyable) = pubkey {
+            Some(pubkeyable.try_into().map_err(|e| e.into())?)
```

Comment:
```suggestion
        let pubkey = if let Some(pubkey) = pubkey {
            Some(pubkey.try_into().map_err(|e| e.into())?)
```
Oops, meant to suggest this during the first round